src/main/java/org/jabref/logic/importer/fetcher/ResearchGate.java

```diff
@@ -0,0 +1,268 @@
+    private final ImportFormatPreferences formatPreferences;
+
+    public ResearchGate(ImportFormatPreferences importFormatPreferences) {
+        this.formatPreferences = importFormatPreferences;
+    }
+
+    /**
+     * Tries to find a fulltext URL for a given BibTex entry.
+     * <p>
+     * Search by title first, as DOI is not searchable directly. When the title is not present, the search is made with DOI via google.com with site:researchgate.net
+     *
+     * @param entry The Bibtex entry
+     * @return The fulltext PDF URL Optional, if found, or an empty Optional if not found.
+     * @throws IOException      if an IO operation has failed
+     * @throws FetcherException if the ResearchGate refuses to serve the page
+     */
+    @Override
+    public Optional<URL> findFullText(BibEntry entry) throws IOException, FetcherException {
+        Objects.requireNonNull(entry);
+
+        // DOI search
+        Optional<String> title = entry.getField(StandardField.TITLE);
+        Optional<DOI> doi = entry.getField(StandardField.DOI).flatMap(DOI::parse);
+
+        // Retrieve PDF link
+        String linkForSearch;
+        Document html;
+        try {
+            if (title.isPresent()) {
+                LOGGER.trace("Search by Title");
+                linkForSearch = getURLByString(title.get());
+                Connection connection = Jsoup.connect(linkForSearch);
+                html = connection
+                        .cookieStore(connection.cookieStore())
+                        .userAgent(URLDownload.USER_AGENT)
+                        .referrer("www.google.com")
+                        .ignoreHttpErrors(true)
+                        .get();
+            } else if (doi.isPresent()) {
+                LOGGER.trace("Search by DOI");
+                // Retrieve PDF link
+                Connection connection = Jsoup.connect(getURLByDoi(doi.get()));
+                html = connection
+                        .cookieStore(connection.cookieStore())
+                        .userAgent(URLDownload.USER_AGENT)
+                        .ignoreHttpErrors(true)
+                        .get();
+            } else {
+                return Optional.empty();
+            }
+        } catch (FetcherException | NullPointerException e) {
+            LOGGER.debug("ResearchGate server is not available");
+            return Optional.empty();
+        }
+        Elements eLink = html.getElementsByTag("section");
+        String link = eLink.select("a[href^=https]").select("a[href$=.pdf]").attr("href");
+        LOGGER.debug("PDF link: {}", link);
+
+        if (link.contains("researchgate.net")) {
+            return Optional.of(new URL(link));
+        }
+        return Optional.empty();
+    }
+
+    String getURLByString(String query) throws IOException, FetcherException, NullPointerException {
```

Comment:
```suggestion
    private String getURLByString(String query) throws IOException, FetcherException, NullPointerException {
```

Reply:
Making it private means I can't call it from the test class... Would "protected" be better?

Reply:
Ah did not see that you have test for it. Then leave it as is